// @virtuals-protocol/game/dist/apiV2.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
class GameClientV2 {
    constructor(apiKey) {
        this.apiKey = apiKey;
        this.runnerUrl = "http://localhost:3001/v1";
        this.client = axios_1.default.create({
            baseURL: this.runnerUrl,
            headers: {
                "Content-Type": "application/json",
                "x-api-key": this.apiKey,
            },
        });
    }
    createMap(workers) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.post("/maps", {
                locations: workers.map((worker) => ({
                    id: worker.id,
                    name: worker.name,
                    description: worker.description,
                })),
            });
            console.log("map", result.data.data);
            return result.data.data;
        });
    }
    createAgent(name, goal, description) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.post("/agents", {
                name,
                goal,
                description,
            });
            console.log("agent", result.data.data);
            return result.data.data;
        });
    }
    getAction(agentId, mapId, worker, gameActionResult, environment, agentState) {
        return __awaiter(this, void 0, void 0, function* () {
            const payload = {
                location: worker.id,
                map_id: mapId,
                environment: environment,
                functions: worker.functions.map((fn) => fn.toJSON()),
                agent_state: agentState,
                version: "v2",
            };
            if (gameActionResult) {
                payload.current_action = gameActionResult;
            }
            const result = yield this.client.post(`/agents/${agentId}/actions`, payload);
            return result.data.data;
        });
    }
    setTask(agentId, task) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.client.post(`/v2/agents/${agentId}/tasks`, {
                task: task,
            });
            return result.data.data.submission_id;
        });
    }
    getTaskAction(agentId, submissionId, worker, gameActionResult, environment) {
        return __awaiter(this, void 0, void 0, function* () {
            const payload = {
                environment: environment,
                functions: worker.functions.map((fn) => fn.toJSON()),
            };
            if (gameActionResult) {
                payload.action_result = gameActionResult;
            }
            const result = yield this.client.post(`/v2/agents/${agentId}/tasks/${submissionId}/next`, payload);
            return result.data.data;
        });
    }
}
exports.default = GameClientV2;
